pkg/subctl/cmd/validate_deployment.go

```diff
@@ -17,46 +17,129 @@ package cmd
+			status.End(cli.Success)
+			return
+		}
+
+		checkPods(item, submariner, OperatorNamespace)
+		checkOverlappingCIDRs(item, submariner)
 	}
 }
 
-func checkPods(config *rest.Config, operatorNamespace string) {
-	submariner := getSubmarinerResource(config)
-	if submariner == nil {
-		status.QueueWarningMessage(submMissingMessage)
-		status.End(cli.Success)
-		return
+func checkOverlappingCIDRs(item restConfig, submariner *v1alpha1.Submariner) {
+	var message string
+	submarinerClient, err := smClientset.NewForConfig(item.config)
+	exitOnError("Unable to get the Submariner client", err)
+
+	localClusterName := submariner.Status.ClusterID
+	endpointList, err := submarinerClient.SubmarinerV1().Endpoints(submariner.Namespace).List(metav1.ListOptions{})
+	if err != nil {
+		exitOnError("Error while listing the endpoints", err)
 	}
 
-	kubeClientSet, err := kubernetes.NewForConfig(config)
+	status.Start("Verifying if cluster CIDRs overlap")
+
+	var localEndpoint submarinerv1.Endpoint
+	overlappingClusters := make(map[string]submarinerv1.Endpoint)
+	for i, source := range endpointList.Items {
+		if localClusterName == source.Spec.ClusterID {
+			localEndpoint = source
+		}
+
+		for _, dest := range endpointList.Items[i+1:] {
+			// Currently we dont support multiple endpoints in a cluster, hence return an error.
+			// When the corresponding support is added, this check needs to be updated.
+			if source.Spec.ClusterID == dest.Spec.ClusterID {
+				message = fmt.Sprintf("Looks like some stale endpoints are present for cluster %q",
+					source.Spec.ClusterID)
+				status.QueueFailureMessage(message)
+				status.End(cli.Failure)
+				break
+			}
+
+			for _, subnet := range dest.Spec.Subnets {
+				overlap, err := util.IsOverlappingCIDR(source.Spec.Subnets, subnet)
+				if err != nil {
+					// Ideally this case will never hit, as the subnets are valid CIDRs
+					message = fmt.Sprintf("Error parsing the CIDR %s in cluster %q", err, dest.Spec.ClusterID)
+					status.QueueFailureMessage(message)
+					status.End(cli.Failure)
+					break
+				}
+
+				if overlap {
+					overlappingClusters[source.Spec.ClusterID] = source
+					overlappingClusters[dest.Spec.ClusterID] = dest
+				}
+			}
+		}
+	}
+
+	if len(overlappingClusters) > 0 {
+		if _, exists := overlappingClusters[localClusterName]; exists {
+			delete(overlappingClusters, localClusterName)
+			message = fmt.Sprintf("localCluster %q with CIDRs %q overlaps with clusters %#v",
+				localClusterName, localEndpoint.Spec.Subnets, overlappingClusters)
+		} else {
+			message = fmt.Sprintf("localCluster %q does not have overlapping CIDRs with other member clusters. "+
+				"However, the following submariner member clusters have overlapping CIDRs %#v", localClusterName,
+				getClusterDetails(overlappingClusters))
+		}
+		status.QueueFailureMessage(message)
+		status.End(cli.Failure)
+	}
+
+	status.QueueSuccessMessage("Clusters do not have overlapping CIDRs")
+	status.End(cli.Success)
```

Comment:
I think we can simplify this code by checking `Clusters` instead of `Endpoints`:

```
        localClusterName := submariner.Status.ClusterID
	
	localCluster, err := submarinerClient.SubmarinerV1().Clusters(submariner.Namespace).Get(localClusterName, metav1.GetOptions{})
	exitOnError("Error getting local Cluster", err)

	localCIDRs := append(localCluster.Spec.ClusterCIDR, localCluster.Spec.ServiceCIDR...)
	status.Start(fmt.Sprintf("Validating CIDRs %v for cluster %q", localCIDRs, localClusterName))

	clusters, err := submarinerClient.SubmarinerV1().Clusters(submariner.Namespace).List(metav1.ListOptions{})
	exitOnError("Error listing Clusters", err)

	result := cli.Success
	for _, c := range clusters.Items {
		if c.Name == localClusterName {
			continue
		}

		for _, cidr := range append(c.Spec.ClusterCIDR, c.Spec.ServiceCIDR...) {
			overlap, err := util.IsOverlappingCIDR(localCIDRs, cidr)
			exitOnError(fmt.Sprintf("Error processing CIDR %Q in cluster %q", cidr, c.Name), err)

			if overlap {
				status.QueueFailureMessage(fmt.Sprintf("CIDR %q in cluster %q overlaps", cidr, c.Name))
				result = cli.Failure
			}
		}
	}

	if result == cli.Success {
		status.QueueSuccessMessage("CIDRs do not overlap with any other cluster")
	}

	status.End(result)
```

Reply:
Thank you @tpantelis. But this code is missing checks for Globalnet CIDRs, Overlapping CIDRs in other clusters, stale endpoints etc. 

With endpoints we have certain advantages as the endpoint.Spec.Subnets is populated based on Globalnet/Vanilla Submariner - so we don't have to explicitly check Globalnet again. Also, an endpoint plays an important role for tunnels, so checking for stale endpoints will help validate any corner cases etc.

Reply:
To cover GN, we can add: `localCIDRs := append(localCIDRs, localCluster.Spec.Global)`. Both the `Cluster` and `Endpoint`  are populated from the env vars via the `SubmarinerSpecification` - the former has the CIDRs in 3 fields and the latter in one.

The purpose of this code is to check for overlapping CIDRs, not stale endpoints, so perhaps we're overloading it and stale endpoints should be another check. 

Overlapping CIDRs in other clusters would be checked when processing those clusters. This code is in the context of a single cluster so I think it should just check that cluster. If every cluster checks every other cluster then we'll have repeated output for overlaps. If we want to check all clusters in one sweep then perhaps we should read from the broker.

Reply:
> To cover GN, we can add: `localCIDRs := append(localCIDRs, localCluster.Spec.Global)`. 

If we keep adding such checks then there will not be much difference between the implementations ;-) 
And on the other hand, AFAICT a stale endpoint has more effect on the cluster functionality compared to a stale cluster object.

> Both the `Cluster` and `Endpoint` are populated from the env vars via the `SubmarinerSpecification` - the former has the CIDRs in 3 fields and the latter in one.

Yes. But it's the latter (aka endpoint.Spec.Subnets) that plays an important role during Tunnel configuration. 

> The purpose of this code is to check for overlapping CIDRs, not stale endpoints, so perhaps we're overloading it and stale endpoints should be another check.

This is basically a validation tool, to point out any issues with the Submariner deployment. The command here is `subctl validate deployment` - so if we validate stale endpoints while checking for overlapping CIDRs, we are optimizing the checks instead of creating new sub-commands. IMHO, we should also look at UI POV to minimize the number of sub-commands instead of creating additional sub-commands to `subctl validate`.

> Overlapping CIDRs in other clusters would be checked when processing those clusters. This code is in the 
> context of a single cluster so I think it should just check that cluster. If every cluster checks every other cluster
> then we'll have repeated output for overlaps. If we want to check all clusters in one sweep then perhaps we 
> should read from the broker.

AFAICT in most cases, users will run this command against a single cluster. Based on my experience supporting users, overlapping clusters is one of the common issues I have seen. If we provide warning about Overlapping Clusters even when it overlaps across other clusters, it saves a lot of time/debugging-effort for our users instead of expecting them to run the `subctl validate deployment` command on each and every cluster to ultimately know that there is some issue with Overlapping CIDRs.

Reply:
ok I agree using the endpoints is better. My main concern is the complexity of the algorithm - it looks to be n^3 and it will parse CIDRs multiple times. This may be an issue with larger deployments but we can always optimize later. Also the overlap messages print a lot of info (entire Endpoint objects) which is a bit hard to read/decipher. Since we're checking for overlap across all clusters anyway, I suggest to simply print specific info for each overlap found and not treat the local cluster any different. This simplifies the code.

```
        status.Start("Checking if cluster CIDRs overlap")

	endpointList, err := submarinerClient.SubmarinerV1().Endpoints(submariner.Namespace).List(metav1.ListOptions{})

	if err != nil {
		status.QueueFailureMessage(fmt.Sprintf("Error listing the Submariner endpoints in cluster %q", localClusterName))
		status.End(cli.Failure)
		return
	}

	for i, source := range endpointList.Items {
		for _, dest := range endpointList.Items[i+1:] {
			// Currently we dont support multiple endpoints in a cluster, hence return an error.
			// When the corresponding support is added, this check needs to be updated.
			if source.Spec.ClusterID == dest.Spec.ClusterID {
				status.QueueFailureMessage(fmt.Sprintf("Found multiple Submariner endpoints (%q and %q) in cluster %q",
					source.Name, dest.Name, source.Spec.ClusterID))
				continue
			}

			for _, subnet := range dest.Spec.Subnets {
				overlap, err := util.IsOverlappingCIDR(source.Spec.Subnets, subnet)
				if err != nil {
					// Ideally this case will never hit, as the subnets are valid CIDRs
					status.QueueFailureMessage(fmt.Sprintf("Error parsing CIDR in cluster %q: %s", dest.Spec.ClusterID, err))
					continue
				}

				if overlap {
					status.QueueFailureMessage(fmt.Sprintf("CIDR %q in cluster %q overlaps with cluster %q (CIDRs: %v)",
						subnet, dest.Spec.ClusterID, source.Spec.ClusterID, source.Spec.Subnets))
				}
			}
		}
	}

	if status.HasFailureMessages() {
		status.End(cli.Failure)
		return
	}

	status.QueueSuccessMessage("Clusters do not have overlapping CIDRs")
	status.End(cli.Success)
```

Note: I added a new method to `Status`:

```
func (s *Status) HasFailureMessages() bool {
	return len(s.failureQueue) > 0
}
```

Reply:
Ok, will update it. Thanks.

Reply:
Sample output when overlapping CIDRs are present:

```
[sgaddam@localhost submariner-operator]$ ./bin/subctl validate deployment 
 ✓ Retrieving Submariner resource from "cluster2"
 ✓ Validating submariner pods in "cluster2"
 ✓ All Submariner pods are up and running
 ✗ Checking if cluster CIDRs overlap
 ✗ CIDR "10.1.0.0/16" in cluster "cluster1" overlaps with cluster "cluster-test" (CIDRs: [100.6.0.0/16 10.1.0.0/16])
```

Sample output in success scenario:
```
[sgaddam@localhost submariner-operator]$ ./bin/subctl validate deployment 
 ✓ Retrieving Submariner resource from "cluster2"
 ✓ Validating submariner pods in "cluster2"
 ✓ All Submariner pods are up and running
 ✓ Checking if cluster CIDRs overlap
 ✓ Clusters do not have overlapping CIDRs
[sgaddam@localhost submariner-operator]$
```